File: server/core/Deck.js
var Deck = class{
	
	constructor(cardList){
		this.cardList = cardList;
	}

	firstStep (){
		return this.getTop();
	}
	
	getTop (){
		var card = this.cardList[0];
		this.cardList.shift();
		return card;
	}
	
	draw (){
		return this.getTop();
	}
	
	damage (){
		return this.getTop();
	}
	
	get (length){
		var card = this.cardList[length];
		this.cardList.splice(length, 1);
		return card;
	}
	
	shuffle (){
		for(var i = this.cardList.length - 1; i > 0; i--){
			var r = Math.floor(Math.random() * (i + 1));
			var tmp = this.cardList[i];
			this.cardList[i] = this.cardList[r];
			this.cardList[r] = tmp;
		}
	}
	
	search (number, mark){
		for(var i = 0; i < this.cardList.length; i++){
			if(this.cardList[i].number == number && this.cardList[i].mark == mark){
				return this.get(i);
			}
		}
		return null;
	}
};

module.exports = Deck;

/*
var Deck = function(cardList){
	this.cardList = cardList;
};

Deck.prototype.firstStep = function(){
	return this.getTop();
};

Deck.prototype.getTop = function(){
	var card = this.cardList[0];
	this.cardList.shift();
	return card;
};

Deck.prototype.draw = function(){
	return this.getTop();
};

Deck.prototype.damage = function(){
	return this.getTop();
};

Deck.prototype.get = function(length){
	var card = this.cardList[length];
	this.cardList.splice(length, 1);
	return card;
};

Deck.prototype.shuffle = function(){
	for(var i = this.cardList.length - 1; i > 0; i--){
		var r = Math.floor(Math.random() * (i + 1));
		var tmp = this.cardList[i];
		this.cardList[i] = this.cardList[r];
		this.cardList[r] = tmp;
	}
};

Deck.prototype.search = function(number, mark){
	for(var i = 0; i < this.cardList.length; i++){
		if(this.cardList[i].number == number && this.cardList[i].mark == mark){
			return this.get(i);
		}
	}
	return null;
};

module.exports = Deck;
 */

File: Python_Server/bp/Deck.py
import random
from .Card import Card
from typing import Union

class Deck:
    def __init__(self, card_list: list):
        """
        デッキ
        :param card_list: カードのリスト
        """
        self.card_list = card_list

    def first_step(self):
        """
        ターンの最初のドロー
        :return: ドローしたカード
        """
        return self.get_top()

    def get_top(self) ->Card:
        """
        デッキの一番上を取得
        :return: 取得したカード
        """
        return self.card_list.pop(len(self.card_list) - 1)

    def draw(self):
        """
        ドロー!
        :return: ドローしたカード
        """
        return self.get_top()

    def damage(self) ->Card:
        """
        ダメージを受けたときにデッキを減らす
        :return: デッキの一番上のカード
        """
        return self.get_top()

    def get(self, length: int) ->Card:
        """
        デッキからカードを一枚取得
        :param length: 何番目のカードか
        :return: 取得したカード
        """
        card = self.card_list[length]
        self.card_list.pop(length)
        return card

    def shuffle(self):
        """
        デッキのシャッフル
        """
        random.shuffle(self.card_list)

    # デッキからカードサーチ
    def search(self, number: int, mark: str) ->Union[Card, None]:
        """
        指定したカードのサーチ
        :param number: 指定するカードの数字
        :param mark: 指定するカードのマーク
        :return: 指定したカードが存在するならそのカードをなければNone
        """
        for i in range(0, len(self.card_list)):
            if self.card_list[i].number == number and self.card_list[i].mark == mark:
                return self.get(i)
        return None
